### PR TITLE
Added module that creates taxonomy terms based on the genres field's value

### PR DESCRIPTION
**Pull request by Rolligta**

### CODE REVIEW

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      }
+
+    }
+
+    foreach ($inputs as $input) {
+
+      $assign = Drupal::entityTypeManager()->getStorage('taxonomy_term')->loadByProperties(['name' => $input, 'vid' => 'genres']);
+      $assign = reset($assign);
+      $assign_id = $assign->id();
+      // $entity->set('field_genre_tags',  $assign_id);
```
Comment (kittiradovics) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]: Please remove don't leave commented code in this file.

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+function genres_to_terms_entity_presave(EntityInterface $entity): void {
+
+  if ($entity->bundle() === 'product') {
+    unset($entity->field_genre_tags);
+    $vid = 'genres';
+    $terms = Drupal::entityTypeManager()->getStorage('taxonomy_term')->loadTree($vid);
+    $inputs = explode(' ', $entity->get('field_genres')->value);
+    $termnames = [];
+    foreach ($terms as $term) {
+
```
Comment (kittiradovics) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]: Please remove the unnecessary empty lines in this file.

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+          'vid' => 'genres',
+
+        ]);
+        $term->save();
+
+      }
+
+    }
+
+    foreach ($inputs as $input) {
```
Comment (kittiradovics) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]: You can merge this foreach with the previous one.

File: composer.json [hunk 4]
```diff
@@ -22,6 +22,7 @@
         "drupal/core-composer-scaffold": "^8.8",
         "drupal/core-project-message": "^8.8",
         "drupal/core-recommended": "^8.8",
+        "drupal/eu_cookie_compliance": "^1.8",
```
Comment (kittiradovics) [inline composer.json]: Shouldn't this be part of a different PR?

**Review by kittiradovics — CHANGES_REQUESTED** [review]

File: composer.lock [hunk 5]
```diff
@@ -2321,6 +2321,101 @@
                 "issues": "https://www.drupal.org/project/issues/ctools"
             }
         },
+        {
+            "name": "drupal/eu_cookie_compliance",
```
Comment (kittiradovics) [inline composer.lock]: These changes should not be added either as you removed the composer.json related changes.

**Review by kittiradovics — CHANGES_REQUESTED** [review]

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\taxonomy\Entity\Term;
+
+/**
+ * Contains hooks and callbacks for genres_to_terms.module.
+ */
+function genres_to_terms_entity_presave(EntityInterface $entity): void {
+  if ($entity->bundle() === 'product') {
+    unset($entity->field_genre_tags);
```
Comment (tamasd) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]: Why?

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    $vid = 'genres';
+    $terms = Drupal::entityTypeManager()->getStorage('taxonomy_term')->loadTree($vid);
+    $inputs = explode(PHP_EOL, $entity->get('field_genres')->value);
+    $termnames = [];
+    foreach ($terms as $term) {
+      $termnames[] = $term->name;
```
Comment (tamasd) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]:
```suggestion
      $termnames[$term->name] = $term;
```

This way you can use `empty()` instead of `in_array()` below

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    foreach ($inputs as $input) {
+      if (!in_array($input, $termnames)) {
```
Comment (tamasd) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]:
```suggestion
      if (empty($termnames[$input])) {
```

See above why

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        $term = Term::create([
+          'name' => $input,
+          'vid' => 'genres',
+        ]);
+        $term->save();
+      }
+      $assign = Drupal::entityTypeManager()->getStorage('taxonomy_term')->loadByProperties(['name' => $input, 'vid' => 'genres']);
```
Comment (tamasd) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]: Why do you do this? After calling `->save()` above, the `$term->id()` will return the created term's id.

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      $assign = reset($assign);
+      $assign_id = $assign->id();
+      $entity->field_genre_tags[] = $assign_id;
```
Comment (tamasd) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]: Instead of adding these one-by-one, collect all ids into an array, and set the field value at the end.

Collecting the term ids in a way that the entity reference field likes it:

```php
$term_ids[] = ['target_id' => $assign_id];
```
After the loop:
```php
$entity->set('field_genre_tags', $term_ids);
```
Make sure that you initialize the `$term_ids` variable with an empty array before the loop starts.

File: web/modules/custom/genres_to_terms/genres_to_terms.module [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+declare(strict_types = 1);
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\taxonomy\Entity\Term;
+
+/**
+ * Contains hooks and callbacks for genres_to_terms.module.
+ */
+function genres_to_terms_entity_presave(EntityInterface $entity): void {
+  if ($entity->bundle() === 'product') {
```
Comment (tamasd) [inline web/modules/custom/genres_to_terms/genres_to_terms.module]:
```suggestion
  if ($entity instanceof NodeInterface && $entity->bundle() === 'product') {
```

**Review by tamasd — CHANGES_REQUESTED** [review]

File: web/modules/custom/genres_to_terms/genres_to_terms.module (hunk 6 above)
Comment (Rolligta) [reply web/modules/custom/genres_to_terms/genres_to_terms.module]: Otherwise if i modify the genre field after creating it(for example replace "Shooter" with Open-World) ,the new taxonomy terms will be added on top of the existing ones,this was the only way i managed to avoid that.

File: web/modules/custom/genres_to_terms/genres_to_terms.module (hunk 9 above)
Comment (Rolligta) [reply web/modules/custom/genres_to_terms/genres_to_terms.module]: because there might be terms that i need to assign,that are already stored in tax vocabulary and wasnt created now

File: web/modules/custom/genres_to_terms/genres_to_terms.module (hunk 6 above)
Comment (tamasd) [reply web/modules/custom/genres_to_terms/genres_to_terms.module]: It is not needed if you overwrite the field's value at the end with the collected term ids.

File: web/modules/custom/genres_to_terms/genres_to_terms.module (hunk 9 above)
Comment (tamasd) [reply web/modules/custom/genres_to_terms/genres_to_terms.module]: But you already loaded those terms in the beginning of the function. Why don't you just get it from `$termnames[$input]`?

File: web/modules/custom/genres_to_terms/genres_to_terms.module (hunk 6 above)
Comment (Rolligta) [reply web/modules/custom/genres_to_terms/genres_to_terms.module]: You"re right,i forgot that the code uses "set" now instead of simply just adding them to it!

**Review by tamasd — APPROVED** [review]

**Review by kittiradovics — APPROVED** [review]
